### PR TITLE
SOL-152741: KTLO: Upgrade MCP Go SDK to v1.7.0 and pin E2E to protocol revision 2025-11-25

### PR DESCRIPTION
**Pull request by helen-zhu-solace**
## Summary

Adds the missing `[Unreleased]` CHANGELOG entry for the MCP Go SDK v1.7.0 bump, moves the E2E protocol pins off `2025-03-26` onto `2025-11-25` behind a single `MCP_PROTOCOL_VERSION`, and makes the pin load-bearing: the handshake now asserts the negotiated revision, and post-initialize requests carry `MCP-Protocol-Version`. Closes out SOL-152741, whose dependency work had already landed incidentally via Dependabot PR #289.

Fixes SOL-152741

## Type of Change
- [x] Documentation update
- [x] Dependency update

## Motivation and Context

SOL-152741 asked for three things: bump `github.com/modelcontextprotocol/go-sdk` to v1.7.0, update the one test whose contract changed, and repoint the E2E scripts at the new protocol revision. The first two already happened — Dependabot PR #289 bumped the SDK on 11 Aug and a follow-up commit in that same PR updated `cmd/server/body_limit_test.go`. Nobody spotted that the ticket covered it, so the E2E pins and the CHANGELOG entry were left behind. This PR finishes both and corrects the record on what the upgrade actually delivered.

### Why the pins go to 2025-11-25, not 2026-07-28

The ticket asked for `2026-07-28`. That revision is not reachable on this server, and pinning it anyway would have been worse than leaving the pins alone.

`2026-07-28` drops the `initialize` handshake in favour of a stateless design — `server/discover` plus per-request `_meta` and mandatory `Mcp-Method` / `Mcp-Name` headers. The SDK gates the entire revision on the stateless transport (`streamable.go:871`):

```go
func (t *StreamableServerTransport) SupportsProtocolVersion(version string) bool {
      if version >= protocolVersion20260728 {
              return t.Stateless && slices.Contains(supportedProtocolVersions, version)
      }
      return slices.Contains(supportedProtocolVersions, version)
}
```

We run stateful — `cmd/server/main.go:1059` passes `nil` options — so `t.Stateless` is false. Probing the real SDK handler, constructed the way `main.go` constructs it:

| Request | Result |
|---|---|
| `initialize` body `2025-11-25`, no header | 200, negotiates **2025-11-25** |
| `initialize` body `2026-07-28`, no header | 200, negotiates **2025-11-25** — silent downgrade |
| `initialize` body `2026-07-28` + `MCP-Protocol-Version` header | **400** `only supported on stateless HTTP servers` |
| `tools/list` + `MCP-Protocol-Version: 2026-07-28` | **400** same |
| `server/discover` at `2026-07-28` | 200, but `supportedVersions` omits `2026-07-28` |

The second row is the trap: setting the pin to `2026-07-28` leaves E2E green while the server quietly answers `2025-11-25`, so the pin would assert a revision that was never exercised. `2025-11-25` is the newest revision this server can actually negotiate, so that is what the scripts now pin, with a comment explaining why so nobody "fixes" it back. Switching to the stateless transport is explicitly out of scope on the ticket and is tracked as its own follow-up.

### Making the pin actually mean something

Two gaps that a pin alone leaves open, both closed here:

- **Nothing checked what was negotiated.** Because the SDK answers a too-new or unrecognized revision with `2025-11-25` rather than an error, a drifted pin runs green while asserting a revision that never appeared on the wire. `mcp_initialize` (and the OAuth suite's `mcp_initialize_as`) now parse `result.protocolVersion` out of the SSE response and fail on mismatch.
- **The pin never reached the transport.** The handshake pin does not steer post-initialize requests: without the `MCP-Protocol-Version` header the transport serves each request at its own default `2025-03-26` (`streamable.go:353-358`, `472-474`). So "E2E exercises what we ship" was only half met even with a correct pin. Every post-initialize request across the three suites now sends the header at `$MCP_PROTOCOL_VERSION`. Verified against a locally-built server: `tools/list` returns 200 with the header at the pin and 400 at `2026-07-28`, matching the stateful gate above.

### One correction worth flagging

The ticket is titled "so the server speaks the 2026-07-28 spec." It does not, and cannot until the stateless decision is made. The bump also did **not** move our protocol ceiling — v1.6.1 and v1.7.0 both cap at `2025-11-25` on a stateful transport:

| SDK | `latestProtocolVersion` | Effective ceiling for us |
|---|---|---|
| v1.5.0 | 2025-11-25 | 2025-11-25 |
| v1.6.1 | 2025-11-25 | 2025-11-25 |
| v1.7.0 | 2026-07-28 | 2025-11-25 (stateful) |

What v1.7.0 bought us is a supported SDK, the improved 413 contract on oversize request bodies, and the new revision's code sitting in the tree ready to switch on. The ticket's summary and implementation notes have been corrected, and its status moved back out of Verifying until this merges.

## Acceptance criteria

| AC | Where it landed | How to verify |
|---|---|---|
| `go.mod` on go-sdk v1.7.0, `go mod tidy` clean | #289 (`407e97f`) | `grep go-sdk go.mod` → `v1.7.0`; `go mod tidy && git diff --exit-code go.mod go.sum` |
| `TestLimitRequestBody_RealSDK_StreamedOversize` asserts the 413 contract, comment matches | #289 | `cmd/server/body_limit_test.go:166-171` |
| Protocol version pinned in the E2E scripts | this PR | `grep -rn MCP_PROTOCOL_VERSION test/` — one definition, three consumers, no `2025-03-26` literal left |
| `make check` green | this PR | build, vet, lint, race tests; coverage 89.3% vs the 85% floor |
| `make e2e-all` green | this PR | 17/17 against two real brokers, handshaking and requesting at `2025-11-25` |
| CHANGELOG entry under `[Unreleased]` | this PR | `CHANGELOG.md` → Changed |
| Every commit carries a DCO `Signed-off-by` | this PR | `git log --format=%B` |

The ticket's starting state says v1.5.0; it was v1.6.1 by the time the work started. The CHANGELOG records the real delta.

## Changes Made

- `test/e2e-common/lib.sh` — `MCP_PROTOCOL_VERSION` (default `2025-11-25`, overridable) as the single source of truth, with the rationale at the definition; new `assert_negotiated_protocol` called from `mcp_initialize`; `mcp_request` and the `initialized` notification send `MCP-Protocol-Version`
- `test/e2e-oauth/helpers.sh` — `mcp_initialize_as` reads the shared variable and asserts the negotiated revision; `mcp_request_as` sends the header
- `test/e2e-llm/run-all.sh` — `precheck` reads the shared variable; `mcp_call` sends the header; the helpers-missing fallback uses the same `export … "${MCP_PROTOCOL_VERSION:-2025-11-25}"` form so an operator override survives that path too
- `docs/internal/e2e-testing.md` — documents the variable, its default, and the silent-downgrade trap; the "all test configuration lives in `.env`" claim narrowed, since suite-wide protocol settings live in `lib.sh`
- `test/e2e-common/README.md` — same variable added to the scaffold table
- `CHANGELOG.md` — `[Unreleased]` → Changed: the 413-vs-400 contract change from the v1.7.0 bump, plus a note that the negotiated revision set is unchanged

No production code changes. `go.mod` is untouched — v1.7.0 is already there.

## Testing

**Test Environment:**
- Go version: go1.25.12 linux/amd64
- OS: Fedora
- Broker version (if applicable): `solace/solace-pubsub-standard:latest` (Dockerized)

**Tests Performed:**
- [ ] Unit tests added/updated
- [ ] Integration tests added/updated (if applicable)
- [x] E2E tests added/updated (if applicable)
- [x] Tested locally with `go test -race ./...`
- [x] Tested with real Solace broker

**Test Coverage:**
- `make check` green — build, vet, lint, race-enabled tests; total coverage 89.3% against the 85% floor
- `make e2e-all` green — 17/17 scenarios (13 standalone, 1 agent, 3 negative-path) against two real brokers, now handshaking *and* issuing requests at `2025-11-25`
- `make e2e-management-all` green — 12/12 config-tool tests; it consumes the edited `e2e-common/lib.sh`
- `make e2e-monitoring-all` green
- `make e2e-oauth-all` green

## Breaking Changes

None.

**Impact:** n/a

**Migration Guide:** n/a

### Code Quality
- [x] Code follows the [coding standards](../.github/CONTRIBUTING.md#coding-standards)
- [x] Self-review completed (checked for typos, logic errors, edge cases)
- [x] Code is well-commented (explains "why", not "what")
- [x] Complex logic has explanatory comments
- [x] No commented-out code or debug statements left in

### Security
- [x] **No credentials in code** (checked all files, commits, and logs)
- [x] Followed [secure logging rules](../docs/internal/secure-logging-rules.md)
- [x] Credential-carrying types implement `slog.LogValuer`
- [x] No security vulnerabilities introduced (checked with golangci-lint/gosec)

### Testing
- [x] All tests pass locally: `go test -race ./...`
- [x] No race conditions detected
- [x] Edge cases covered by tests
- [x] Error paths tested
- [x] Test names clearly describe what is being tested

### Documentation
- [ ] README.md updated (if user-facing changes)
- [x] docs/ updated (if architecture/design changes)
- [ ] godoc comments added/updated for exported functions
- [x] CHANGELOG.md updated (if applicable)
- [ ] Examples updated (if applicable)

### Git & CI
- [x] All commits are signed off (DCO: `git commit -s`)
- [x] Commits have clear, descriptive messages
- [x] Branch is up to date with main (rebased if needed)
- [x] No merge conflicts
- [x] CI checks passing (lint, build, test, E2E)

### Breaking Changes (if applicable)
- [ ] Breaking changes documented in commit message
- [ ] Breaking changes documented in PR description
- [ ] Migration guide provided
- [ ] Deprecated functionality marked with comments and deprecation notices

## Screenshots (if applicable)

n/a

## Additional Notes

Two follow-ups the ticket lists, neither touched here.

**Session posture / stateless transport.** `deploy/kubernetes/deployment.yaml:8` pins `replicas: 1` with no session affinity — scaling past one replica today returns `404 session not found` to any client whose request hops instances. Going stateless fixes that and unlocks `2026-07-28`. The cost looks low on paper: we make no server-to-client calls (the only `Notify` in the tree is `signal.Notify` for SIGTERM) and E2E only POSTs, so the features stateless mode gives up are ones we do not use. The open risks are client compatibility — clients expecting a session ID or the GET/SSE stream — and the E2E rework, since all four suites read `Mcp-Session-Id` and hard-fail without it. This needs its own ticket; the `lib.sh` comment points at "no ticket yet" rather than at SOL-152741, which closes with this PR.

**DCR to CIMD.** `docs/authentication.md` documents Dynamic Client Registration; `2026-07-28` deprecates it in favour of Client ID Metadata Documents. Earliest removal is a revision on or after 28 July 2027, so no urgency.

## Related Issues/PRs

- Related to #289

### CODE REVIEW

File: test/e2e-oauth/helpers.sh [hunk 1]
```diff
@@ -533,7 +533,8 @@ restart_oauth_server() {
 # ── MCP wire (token-parameterized) ──────────────────────────────────────────
 # Mirrors e2e-common/lib.sh's mcp_initialize/mcp_call_tool/mcp_request, taking
 # the bearer token as an argument instead of the shared static dev token, and
-# using --cacert for the server's self-signed cert.
+# using --cacert for the server's self-signed cert. See mcp_initialize there for
+# why protocolVersion pins 2025-11-25 rather than go-sdk's latest revision.
```
[the remark this answers is not in the document]
Comment (helen-zhu-solace) [reply test/e2e-oauth/helpers.sh]: Resolved in [0d7a945](https://github.com/SolaceProducts/solace-broker-mcp/pull/310/commits/0d7a945c23d44cadb9a59d990e465928950c0f45)

File: test/e2e-llm/run-all.sh [hunk 2]
```diff
@@ -67,10 +67,12 @@ mcp_call() {
         -X POST "$MCP_URL/mcp" -d "$body" 2>/dev/null
 }
 
+# protocolVersion pins 2025-11-25, the newest revision this server negotiates;
+# see mcp_initialize in ../e2e-common/lib.sh for why it is not go-sdk's latest.
```
[the remark this answers is not in the document]
Comment (helen-zhu-solace) [reply test/e2e-llm/run-all.sh]: Resolved in [0d7a945](https://github.com/SolaceProducts/solace-broker-mcp/pull/310/commits/0d7a945c23d44cadb9a59d990e465928950c0f45)

File: test/e2e-common/lib.sh [hunk 3]
```diff
@@ -579,7 +588,7 @@ mcp_initialize() {
             "id": 1,
             "method": "initialize",
             "params": {
-                "protocolVersion": "2025-03-26",
+                "protocolVersion": "2025-11-25",
```
[the remark this answers is not in the document]
Comment (helen-zhu-solace) [reply test/e2e-common/lib.sh]: Resolved in [0d7a945](https://github.com/SolaceProducts/solace-broker-mcp/pull/310/commits/0d7a945c23d44cadb9a59d990e465928950c0f45)

**Comment by bczoma** [comment]
## Review: E2E protocol pin + go-sdk v1.7.0 CHANGELOG

**Jira:** `SOL-152741` (status: Verifying)

### Summary

A small, well-argued cleanup: it retires the `2025-03-26` protocol pin duplicated across three E2E `initialize` handshakes in favour of a single `MCP_PROTOCOL_VERSION` in `e2e-common/lib.sh`, and adds the `[Unreleased]` CHANGELOG entry that should have accompanied the go-sdk v1.7.0 bump when it landed via #289.

The technical judgment is sound and I verified it independently against SDK source rather than taking the PR body's word: `2026-07-28` really is gated on the stateless transport, the server really is stateful, and pinning the ticket's requested revision would have left E2E green while the SDK silently negotiated back down. `go.mod` is on v1.7.0 and the 413 canary test genuinely asserts the new contract on this branch.

What's left is presentation and hygiene, not correctness. Nothing here blocks merge on risk grounds.

---

### Issues

#### 🟡 Title asserts what the PR itself disproves — PR-level

The title is copied verbatim from the Jira summary: "…so the server speaks the 2026-07-28 spec." The body then spends three paragraphs arguing the server neither does nor can, on a stateful transport. The title is what survives into `git log`, release notes, and search — a merge commit asserting the server speaks 2026-07-28 will mislead the next person debugging protocol negotiation.

**Fix:** retitle to what the PR does, e.g. `SOL-152741: Pin E2E protocol version to 2025-11-25 and record the go-sdk v1.7.0 bump in CHANGELOG`. The Jira summary is wrong at the source and should be corrected too.

#### 🟡 Ticket deviation lives only in the PR body — PR-level

The ticket asks for the pin to go to `2026-07-28`; this deliberately pins `2025-11-25`. The reasoning is correct and durable — it's committed as a comment in `lib.sh`, not just prose — but SOL-152741 has no record of it, and the ticket sits in **Verifying** while the PR is still open and blocked. An auditor reading the ticket later sees a requested change that wasn't applied, with no trace of why.

**Fix:** post the rationale as a Jira comment, correct the summary and implementation notes, and move the ticket back out of Verifying until this merges.

#### 🟡 Half the ticket's ACs are unverifiable from this diff — PR-level

Two of six ACs — `go.mod` on v1.7.0 with clean tidy, and the `TestLimitRequestBody_RealSDK_StreamedOversize` 413 assertion — landed in #289, not here. The body says so in prose, but there's no AC-by-AC mapping, so a reviewer signing off on SOL-152741 has to rediscover that. Both are in fact satisfied on this branch: `go.mod:9` is at v1.7.0 via 407e97f, and `body_limit_test.go:166-171` asserts 413 with a comment naming the v1.7.0 change.

**Fix:** add a short AC table to the body — each criterion, where it landed (this PR vs. #289 SHA), how to verify.

#### 🟡 The de-duplication leaves one literal behind — `test/e2e-llm/run-all.sh:35`

The fallback is `MCP_PROTOCOL_VERSION='2025-11-25'` — unconditional, unexported, hardcoded — against `lib.sh`'s `export MCP_PROTOCOL_VERSION="${MCP_PROTOCOL_VERSION:-2025-11-25}"`. Two deltas: an operator override is honoured on the helpers-present path but silently clobbered on the helpers-missing path, and the value isn't exported to children there. The second literal is guarded only by a comment asking a human to keep it in step — a comment isn't a mechanism, and this is the exact drift the PR set out to remove.

**Fix:** use the identical form (`export MCP_PROTOCOL_VERSION="${MCP_PROTOCOL_VERSION:-2025-11-25}"`), or drop the fallback and let `set -u` fail loudly — the suite can't run without the helpers' logging and colour functions anyway.

#### 🟡 New E2E knob is documented nowhere — `docs/internal/e2e-testing.md:9`, `:70–94`

`MCP_PROTOCOL_VERSION` is now externally overridable, but §Shared scaffold (line 9) enumerates what `lib.sh` provides and omits it, and §Configuration (line 72) asserts "All test configuration lives in a single file: `test/e2e-basic-mcp/.env` … the source of truth for ports, credentials, and MCP server env vars" — now false for this knob. `test/e2e-common/README.md` doesn't mention it either.

**Fix:** one line after line 9 naming the variable, its default, and the trap the code comment documents (raising the pin doesn't fail — the SDK negotiates back down silently). Optionally soften line 72 to note that suite-wide protocol settings live in `lib.sh`.

#### 🟢 CHANGELOG overclaims on `server/discover` — `CHANGELOG.md:21`

The entry says `2026-07-28` is "neither advertised in a `server/discover` response nor accepted on a request." The first half is right. The second isn't: `go-sdk@v1.7.0/mcp/streamable.go:1516` explicitly exempts discover from the stateful rejection — `if !c.stateless && jreq.Method != methodDiscover` — with a comment saying discover "should learn about the supported protocols from the DiscoverResult response." So a `server/discover` at `2026-07-28` *is* accepted; it just returns a `SupportedVersions` list that excludes it. A reader integrating a client would expect a 400 on the one method whose success is the intended probe path.

**Fix:** "…so `2026-07-28` is not advertised in the `server/discover` response, and any other request carrying it is rejected with `400` — adopting it remains a separate decision."

#### 🟢 Version mismatch between PR and ticket — `CHANGELOG.md`

The entry says v1.6.1 → v1.7.0; the ticket says v1.5.0 → v1.7.0. The PR is right (407e97f bumped 1.6.1→1.7.0) — the ticket is stale. No change here; fix the ticket.

#### 💡 CHANGELOG entry is ~230 words, half of it documenting a non-change — `CHANGELOG.md:21`

`RELEASING.md` promotes this block verbatim at tag time. The 413-vs-400 half records an actual delta; the protocol half explains what did *not* change, and that rationale already lives better-placed in `lib.sh:570-581` and the PR body. Consider reducing the second half to one sentence.

#### 💡 The pin is a request-side hint, not a gate — `test/e2e-common/lib.sh:582`

`mcp_initialize` never checks `result.protocolVersion` in the response. I initially filed this as Major on the theory that a stray override or drift would produce a green run against an unintended revision. On measuring it, that doesn't hold:

- **No trigger exists.** `grep -rn MCP_PROTOCOL_VERSION` returns only the six lines this PR creates; `grep -rn PROTOCOL .github/` returns nothing. The name is new here, so no pre-existing setter can collide. The `run-all.sh:35` duplicate sits in the `else` branch, taken only when `test/e2e-llm/helpers.sh` is absent — CI runs from a full checkout, so it never executes there.
- **The compensating control is structural.** `go-sdk@v1.7.0/mcp/shared.go:68-79`: `negotiatedVersion` returns the client string only if supported and `< 2026-07-28`, **otherwise it returns `2025-11-25` — the pinned value itself.** A typo, garbage, or an over-ambitious `2026-07-28` all land on the intended revision. Only the three valid *older* strings negotiate differently.
- **Blast radius is nil.** No revision-conditional logic in `internal/` or `cmd/`. In the SDK the stored `ProtocolVersion` is read in one place (`server.go:1544`, gated on `>= 2026-07-28`, unreachable here); transport branches key off the `MCP-Protocol-Version` *header*, which the scripts never send. Evidence: the pin was `2025-03-26` before this PR and every suite was green, because nothing depends on it.

Still worth the two lines as hygiene — it documents intent and would catch a future deliberate override to a valid older revision. The SDK does echo the negotiated version (`server.go:1985`), so **parse `result.protocolVersion` and fail on mismatch**.

#### 💡 The pin doesn't steer the transport at all — `test/e2e-common/lib.sh`

More material than the above: because no suite sends the `MCP-Protocol-Version` header, the transport serves every post-initialize request at its default `2025-03-26` (`streamable.go:353-358`, `472-474`) regardless of the handshake pin. So "E2E exercises what we ship" is only partially met even with a correct pin *and* a correct assertion. Worth sending the header on post-initialize requests if that's the goal.

#### 💡 Comment points at a ticket this PR closes — `test/e2e-common/lib.sh:581`

"Revisit alongside the stateless-transport decision (SOL-152741)" becomes a dead pointer once SOL-152741 closes. Reference the stateless-transport follow-up, or say "no ticket yet."

#### 💡 Fixup commit in history — PR-level

Two commits, the second `SOL-152741: Address Copilot's comments`. Both carry valid DCO trailers. Squash-merge or squash locally.

---

### Strengths

- **The description does the hard part.** It doesn't just say what changed — it explains why the obvious change (pinning `2026-07-28`) would have been actively harmful, names the SDK gating function, and backs it with a probe table of real handler responses.
- **Intellectually honest about the ticket.** It flags that the ticket title is factually wrong and that the bump didn't move the protocol ceiling, rather than quietly shipping something that resembles the ask.
- **The rationale is durable.** It lives as a 13-line comment at the value in `lib.sh`, not only in the PR body, so whoever next wants to raise the pin finds the reason at the site. It correctly names the silent-downgrade trap and doesn't overclaim.
- **Every load-bearing CHANGELOG claim except the discover clause verifies against source** — the 413 branch on `*http.MaxBytesError`, the `limitRequestBody` short-circuit at `main.go:399-402`, the unchanged revision set, the `nil` options at `main.go:1059`.
- **Shell correctness is clean.** Quoting is right at all three sites, the variable is in scope on every sourcing path, `${…:-}` closes the empty-value hazard, and no `2025-03-26` literal survives anywhere in the tree.
- **All three Copilot comments were substantively addressed**, not just replied to.

---

### Verdict

| Category | Assessment |
|---|---|
| Correctness | ✅ |
| Code Quality | ⚠️ |
| Test Coverage | ✅ |
| Architecture | — not reviewed (no production code or design surface in the diff) |
| Security | — not reviewed (no auth, crypto, input-validation, or dependency surface in the diff) |
| Documentation | ⚠️ |
| PR Scope & Description | ⚠️ |

**Recommendation: 🔄 Approve with minor changes.** No blocker survived scrutiny. The remaining items are all cheap: retitle the PR, amend the ticket, align the `run-all.sh` fallback, document the env var, and fix the discover clause in the CHANGELOG.

**Review by solace-aross — DISMISSED** [review]
Approving. Verified the central technical claim against the actual SDK source rather than the PR's prose, since that claim is the whole justification for pinning at 2025-11-25 instead of the ticket's 2026-07-28.

Checked against `go-sdk@v1.7.0` in the module cache:
- `main.go:1059` passes `nil` options to `NewStreamableHTTPHandler`. In the SDK, nil opts leaves `Stateless` at its zero value, `false` (`streamable.go:237-239`).
- `SupportsProtocolVersion` (`streamable.go:871-876`) gates `>= 2026-07-28` on `t.Stateless` exactly as quoted in the PR description, word for word.
- The CHANGELOG's 413-vs-400 claim also checks out: diffing `v1.6.1` against `v1.7.0`, the old `servePOST` had one unconditional `http.Error(w, "failed to read body", 400)`. The new one added `var mbe *http.MaxBytesError; if errors.As(err, &mbe) { http.Error(w, fmt.Sprintf("request body exceeds %d bytes", mbe.Limit), 413) }` (`streamable.go:1418-1424`) ahead of the same 400 fallback. Matches the CHANGELOG entry's wording exactly.

So the pin at 2025-11-25 is correct: raising it without the stateless-transport decision would silently negotiate back down rather than fail, which is worse than the current pin. Good catch, and good that it's written down as a comment in `lib.sh` so it doesn't get "fixed" back later.

Copilot's three comments (literal duplication risk, one stale-comment risk, one ambiguous cross-reference) are all resolved at this head via the `MCP_PROTOCOL_VERSION` env var centralized in `lib.sh` and sourced by the other two scripts.

No production code touched, CI green across all five e2e suites plus the full gate. Nothing here blocks merge.

**Review by solace-lootii — DISMISSED** [review]
LGTM 

Went through the diff and cross-checked the claims against the codebase (not just the PR description) — solid work.

**Scope check:** 4 files, +27/-4, no production code touched. `go.mod` is correctly left alone since v1.7.0 already landed via #289 — good instinct not to re-touch it.

**Verified:**
- CHANGELOG's 413-vs-400 claim matches `limitRequestBody`/`body_limit_test.go` — accurate, not just narrated.
- The `MCP_PROTOCOL_VERSION` sourcing chain actually works: `e2e-oauth/helpers.sh` and `e2e-llm/helpers.sh` (transitively via `e2e-monitoring/helpers.sh`) both source `e2e-common/lib.sh`, so the single `export MCP_PROTOCOL_VERSION="${MCP_PROTOCOL_VERSION:-2025-11-25}"` default reaches every consumer, and it's declared before first use.
- The justification for pinning `2025-11-25` instead of the ticket's literal `2026-07-28` ask holds up: that revision requires the SDK's stateless transport, this server runs stateful, and the SDK silently downgrades rather than erroring — so pinning `2026-07-28` would've asserted an untested path. Good call pushing back on the ticket instead of cargo-culting the version string, and it's disclosed prominently rather than buried in a footnote.
- The two declined follow-ups (stateless-transport migration, DCR→CIMD) are explicitly scoped out with reasoning, not silently dropped.

**One nit (non-blocking):** [`test/e2e-llm/run-all.sh#L35`](../blob/main/test/e2e-llm/run-all.sh#L35) hardcodes `MCP_PROTOCOL_VERSION='2025-11-25'` as a fallback for when `helpers.sh` is missing from the checkout. It's kept in sync with `lib.sh`'s default only by a comment, not a shared constant — if the canonical pin ever moves, this copy could go stale in exactly the degraded CI environment nobody's watching. Low severity (rare code path, already self-documented) — fine to punt to a follow-up rather than block on it.

Approving.

**Review by bczoma — APPROVED** [review]
Approving — `1bfb4ef` addresses everything raised.

The two items I'd measured as low-risk nits you took anyway, and they were the right ones to take: `assert_negotiated_protocol` turns the pin into a gate, and sending `MCP-Protocol-Version` on post-initialize requests closes the half that actually mattered — without it the transport was serving every request at its own `2025-03-26` default no matter what the handshake negotiated, so "E2E exercises what we ship" was only half true even with a correct pin. That's now genuine rather than nominal.

Also confirmed: the retitle, the AC table mapping the two #289-landed criteria with verify commands, the `run-all.sh` fallback in `lib.sh`'s exact form so an operator override survives that path, `MCP_PROTOCOL_VERSION` documented in both `docs/internal/e2e-testing.md` and `test/e2e-common/README.md` (including the line 72 correction), the dead SOL-152741 pointer replaced, and the `server/discover` clause in the CHANGELOG fixed.

One thing I looked at and decided against asking for: `precheck()` in `test/e2e-llm/run-all.sh:81` is the only initialize path that doesn't call `assert_negotiated_protocol`. Not worth a round — the drift it would catch already fails four other suites, `mcp_call` in that file does send the header so the suite runs at the pinned revision regardless, and e2e-llm is the daily job rather than a merge gate. Fold it in if you push again for another reason; otherwise leave it.

Non-blocking, for whenever: SOL-152741 still carries the original summary ("so the server speaks the 2026-07-28 spec") and sits in Verifying. The pin rationale deserves a comment on the ticket so the deviation from the requested `2026-07-28` isn't only recorded here.
